src/Linker.cs

```diff
@@ -0,0 +1,382 @@
+                var nameBytes = Encoding.UTF8.GetBytes(name);
+                fixed (byte* modulePtr = moduleBytes, namePtr = nameBytes)
+                {
+                    var error = Native.wasmtime_linker_define(handle, modulePtr, (UIntPtr)moduleBytes.Length, namePtr, (UIntPtr)nameBytes.Length, ext);
+                    if (error != IntPtr.Zero)
+                    {
+                        throw WasmtimeException.FromOwnedError(error);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Defines WASI functions in the linker.
+        /// </summary>
+        /// <remarks>
+        /// When WASI functions are defined in the linker, a store must be configured with a WASI
+        /// configuration.
+        /// </remarks>
+        public void DefineWasi()
+        {
+            var error = Native.wasmtime_linker_define_wasi(handle);
+            if (error != IntPtr.Zero)
+            {
+                throw WasmtimeException.FromOwnedError(error);
+            }
+        }
+
+        /// <summary>
+        /// Defines an instance with the specified name in the linker.
+        /// </summary>
+        /// <param name="context">The store context that owns the instance.</param>
+        /// <param name="name">The name of the instance to define.</param>
+        /// <param name="instance">The instance to define.</param>
+        public void DefineInstance(StoreContext context, string name, Instance instance)
+        {
+            if (name is null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (instance is null)
+            {
+                throw new ArgumentNullException(nameof(instance));
+            }
+
+            unsafe
+            {
+                var nameBytes = Encoding.UTF8.GetBytes(name);
+                fixed (byte* namePtr = nameBytes)
+                {
+                    var error = Native.wasmtime_linker_define_instance(handle, context.handle, namePtr, (UIntPtr)nameBytes.Length, instance.instance);
+                    if (error != IntPtr.Zero)
+                    {
+                        throw WasmtimeException.FromOwnedError(error);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Instantiates a module with imports from items defined in the linker.
+        /// </summary>
+        /// <param name="context">The store context to instantiate in.</param>
+        /// <param name="module">The module to instantiate.</param>
+        /// <returns>Returns the new instance.</returns>
+        public Instance Instantiate(StoreContext context, Module module)
+        {
+            if (module is null)
+            {
+                throw new ArgumentNullException(nameof(module));
+            }
+
+            var error = Native.wasmtime_linker_instantiate(handle, context.handle, module.NativeHandle, out var instance, out var trap);
```

Comment:
This is a point where wasm code runs so in other bindings I typically expect it to have roughly the same "guards" as `Func::call` (in w/e form that has). One thing this might be missing is handling of the exception propagation? 

(I'm just assuming that all C# exceptions are caught, converted to a trap, and then when the trap is received on the other side we realize that  the C# exception should be re-raised. Assuming that unwinding via C# through Rust code is unsound)

Reply:
Any calls from wasm to .NET will catch .NET exceptions at the entry site and return a trap back as the result of the host function callback.

There shouldn't be any (catchable) .NET exceptions propagating out from a call to `wasmtime_linker_instantiate`; we check that the returned trap is non-zero and then throw the trap exception.

Reply:
Opened issue #63 for this.